service/dap/server.go

```diff
@@ -419,7 +420,8 @@ func (s *Server) onInitializeRequest(request *dap.InitializeRequest) {
 	response.Body.SupportsTerminateRequest = false
 	response.Body.SupportsRestartRequest = false
 	response.Body.SupportsFunctionBreakpoints = false
-	response.Body.SupportsStepBack = false
+	// TODO(lggomez): research nehalem and zen processors detection for rr as they are the only supported platforms
+	response.Body.SupportsStepBack = runtime.GOOS == "linux"
```

Comment:
This is an interesting question. If rr is not installed, I'd prefer the user to get an error on the stepback actions or show them a warning or some sorts rather that not showing any UI at all like in this case (which is because the only tool supporting this functionality isn't supported by the host). cc @hyangah

Reply:
I think enabling unsupported optional capabilities is misleading when there is a clear mechanism to avoid that. Are you hoping the error will inform them that they need rr installed? Have you actually tried triggering that error to see how informative it would be?

Reply:
@polinasok I'll get into that tomorrow and come back with results, I have a windows machine to test this. The major change to the service is that onInitializeRequest will return an error on unsupported platforms/scenarios like this one (so, how informative the message is can be up to us)

Reply:
Hmm, I am not sure I understand. What about Windows? Isn't rr only available on Linux? 
In any case, I thought we were discussing the following two options:

1. if linux machine with rr => set capability to true
otherwise (on mac, windows and linux without rr), set capability to false
That means:
  -- rr installed + rr backend/mode in launch request + step back clicked = success
  -- rr installed + not rr backend in launch request + sttep back clicked = error
  -- no rr installed + rr backend/mode in launch request = error
  -- no rr installed + not rr backend/mode in launch request  = UI has no step back

2.  if linux machine => set capability to true
otherwise (on mac and windows), set to false
  -- Linux + rr installed  - same as above
  -- Linux + no rr + rr backend/mode in launch request = error
  -- Linux + no rr + not rr backend/mode in launch request = UI has step back button, gives error if clicked
  -- not Linux + rr backend/mode in launch request = error
  -- not Linux + not rr backend/mode in launch request  = UI has no step back

Neither case produces any errors at the time of the initialize request.

Reply:
I was thinking about this more. Were you by any chance considering errors in initialize request because --backend could be supplied on the command-line to solve the timing problem with the capabilities not matching the user mode that enables the controls?

Reply:
The current code is outdated since I added the full platform check but the gist of it is that yes, I'm experimenting with how an error from  the initialize request would behave. Locally I found out that it doesn't show any messages (even if the error response body has showUser = true). So it looks like a bug or additional support lacking on the vs code adapter

I won't be returning an error at this stage after all I guess but I want to check how the user interaction would be. If the platform does not support it, the service will set SupportsStepBack and that's it (this is still not pushed). On the other side, if SupportsStepBack is supported but rr is not installed, I'll let the service fail on the service calls that require it and propagate the error to the user (since this is being done anyways)

Reply:
Did you happen to test whether setting SupportsStepBack in a [capabilities event](https://microsoft.github.io/debug-adapter-protocol/specification#Events_Capabilities) works? Because that would be the correct way to do this. I'm not too keen on adding a bunch of CPUID checks to `service/dap` that don't actually work anyway because the user doesn't end up requesting the `rr` backend.

Reply:
I looked at this myself. Doing this:

```
s.send(&dap.CapabilitiesEvent{ *newEvent("capabilities"), dap.CapabilitiesEventBody{ dap.Capabilities{ SupportsStepBack: true } } })
```

inside onLaunchRequest enables the step back controls. All the code doing the cpuid checks should be reverted and instead do this if the requested backend is rr.

Reply:
@polinasok @aarzilli I've updated the main comment with a more detailed description of the current state of the PR and the changes made. As mentioned on it, the backend is not sent anymore with replay mode and now it is inferred by the `onLaunchRequest` handlers. 

Having a host platform compatibility check would be a nice to have IMO but if you both agree that having the third party dependency and logic on such a low level just to check if the tool is runnable is going to be a burden on the debugger then I will resort to the CapabilitiesEvent route instead

Reply:
The problem it that the check doesn't do anything. Manifesting the stepback UI always on supported architectures is wrong because it doesn't work unless the user selects the `rr` backend and thay can always do that regardless of the amout of checking we do and if they do and it isn't supported they are going to get an error no matter what. On top of that both intel/amd could make new cpus that fail the check but are supported and rr could expand its support to other cpus (like it did recently) and I don't want to have to keep up with them.

Reply:
Point taken. I've updated the implementation with the CapabilitiesEvent responses, so it should be much cleaner now from this viewpoint